Update maintainer and CI shell scripts

jngrad opened the pull request:
Fixes #3242

Description of changes:
- reduce number of shellcheck warnings from 238 down to 104
   - rules: `SC2002 SC2006 SC2061 SC2063 SC2086 SC2116 SC2148 SC2166 SC2196`
   - 50 of the remaining warnings are false positives from `build_cmake.sh` (`SC2086 SC2154`)
- refactor outdated maintainer shell scripts
- remove deprecated shell syntax and shell commands
- use POSIX alternatives to bashisms when the change is trivial
- quote string variables to guard against whitespace-related issues
- start unit tests before integration tests
- run unit tests for all operating systems (Debian, OpenSUSE, Fedora)

Notes:
- `shellcheck` is unsuitable for CI in its current form (e.g. no rule whitelisting capability)
- the shell scripts now have an homogeneous syntax and few linter warnings, which is desirable if we decide to port them to Python in the future

jngrad committed:
CI: run unit tests before integration tests

jngrad committed:
Use || instead of -o

POSIX does not specify test for > 4 arguments:
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/test.html

shellcheck rule: SC2166

jngrad committed:
Use new-style command substitutions

POSIX allows both the legacy `cmd` and the new-style $(cmd) syntaxes:
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_03
However, POSIX recommends using the new-style syntax, as it has less
restrictions and allows nesting without escape characters:
https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xcu_chap02.html#tag_23_02_06_03

shellcheck rule: SC2006

jngrad committed:
Add braces around shell variables

jngrad committed:
Double quote string variables

shellcheck rule: SC2086

jngrad committed:
Formatting

jngrad committed:
Fix script logic

The original pattern wouldn't work if a filename was the suffix of
another, e.g. `lb.py` would be marked as used in the CMakeLists.txt
because of a match at the line `python_test(FILE engine_lb.py)`.
The new pattern is also not a regular expression: the dot in the
file extension doesn't match any character, only the dot character.

jngrad committed:
Use POSIX grep

Replace deprecated `egrep` by `grep -E`.

shellcheck rule: SC2196

jngrad committed:
Remove inexistent files from patterns

jngrad committed:
Add new files in patterns

jngrad committed:
Fix broken stdout

jngrad committed:
Fix broken script

jngrad committed:
Update shebangs

Add a shebang in all shell scripts. Downgrade shebangs from Bash to
Dash (sh) in small scripts that don't use Bash-specific syntax. Use
the most portable shebang (`#!/usr/bin/env <shell>`) to include the
user $PATH in the search path.

shellcheck rule: SC2148

jngrad committed:
Remove superfluous `echo` and `cat`

shellcheck rules: SC2002 SC2116

jngrad committed:
Properly escape globs in `grep` and `find`

shellcheck rules: SC2061 SC2063

jngrad committed:
Run scripts in root directory

jngrad committed:
Document shell scripts

jngrad committed:
Add copyright headers

jngrad committed:
Run unit tests on all operating systems

Separate unit tests from python tests. Run unit tests on Fedora,
Debian and OpenSUSE.

fweik commented:
Is it possible to run the scripts in the CI in POSIX compatibility mode (e.g. using `sh`, and not `bash`)?
This would be a simple test that they actually are POSIX compatible.

jngrad commented:
Not really. For example `build_cmake.sh` has 12 bashisms. I've already downgraded scripts that could run in dash with trivial changes. For the less trivial bashisms, I could have a look if you want, but some of the known POSIX equivalents (see [SC2039](https://github.com/koalaman/shellcheck/wiki/SC2039)) don't look very maintainable.

jngrad committed:
Use POSIX syntax

Reduce bashisms. Downgrade shebangs from Bash to Dash (sh) in
POSIX-compliant shell scripts. Use the most portable shebang
(`#!/usr/bin/env <shell>`) to include the user $PATH in the
search path.

jngrad commented:
I've removed more bashisms, but several shell scripts still rely on useful bash-specific features:
* variable indirections don't seem to have a POSIX equivalent without `eval`
* shell traps: SIGINT, SIGTERM and SIGQUIT automatically trigger SIGEXIT in bash, while in POSIX one has to write a custom trap for each one of them (e.g. a SIGINT trap must forward SIGINT)
* `$BASH_COMMAND` and `$LINENO` don't exist in POSIX (useful for debug messages)
* arrays only exist natively in bash
* `build_cmake.sh`  downloads and executes a codecov script which is not POSIX-compliant

List of bash scripts:
```sh
head -n 1 $(find maintainer/ testsuite/ -name "*.sh")
```
```
maintainer/CI/build_cmake.sh
maintainer/CI/build_docker.sh
maintainer/CI/doc_warnings.sh
maintainer/benchmarks/runner.sh
maintainer/benchmarks/suite.sh
testsuite/cmake/BashUnitTests.sh
testsuite/cmake/test_install.sh
```

jngrad committed:
Quote string variables

fweik commented:
I also don't think it is so super important. At the end of the day most of those scripts only run in an environment that we control.

jngrad committed:
Quote string variables

mkuron commented on maintainer/CI/dox_warnings.sh:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/usr/bin/env sh
```
`/usr/bin/env sh` is useless as `/usr/bin/env` is not guaranteed by POSIX (neither is `/bin/sh`, by the way). I would just use `/bin/sh`.

jngrad replied:
Here the idea was to be more portable, as the program `env` reads the user `$PATH`. I put this change in the POSIX bucket for convenience, since most scripts modified in that commit had heterogenous shebangs that needed fixing, although it has nothing to do with POSIX. In fact, `sh` might not even point to the `dash` interpreter.

mkuron reviewed (approved):
Looks fine. We should really migrate this mess to Python scripts ASAP though.

KaiSzuttor reviewed (approved):
LGTM, it seems like we develop a DevOps library in bash without knowing it :smile:

jngrad commented:
bors r=KaiSzuttor
